sstables/generation_type.hh

```diff
@@ -36,6 +38,22 @@ constexpr int64_t generation_value(generation_type generation) {
     return generation.value();
 }
 
+class generation_generator {
+public:
+    virtual generation_type generate() = 0;
+    virtual ~generation_generator() = default;
+};
+
+class int_generation_generator final : public generation_generator {
+    std::atomic<int64_t> _last_generation;
```

Comment:
Why atomic?

We may need to keep the last int generation even if it's not in use, if we allow configuring the generation type back to int.

Reply:
do we really support changing the generation id back to int? thought there is no u turn here.

Reply:
> > Why atomic?
> 
> good question. i asked this to myself before copying it from the existing code. just wanted to preserve the existing implementation as much as possible, and focus on the generalizing it with a generator class, and keep the cleanups in a separate commit(s).
> 

explained in the comment in code. in short, this is due to how reshard is implemented -- we generate generation identifier from another shard on behalf of the owner shard of a given sstable.